### PR TITLE
MacOS warnings

### PR DESCRIPTION
**Pull request by andyp1per**
Nuke clang warnings about fabsf()

### CODE REVIEW

File: libraries/AP_Math/definitions.h [hunk 1]
```diff
@@ -117,3 +117,30 @@ static const double WGS84_E = (sqrt(2 * WGS84_F - WGS84_F * WGS84_F));
...
+inline double safe_fabs(const double v) {
+#ifdef ALLOW_DOUBLE_MATH_FUNCTIONS
+    return fabs(v);
+#else
+    return fabsf(static_cast<float>(v));
+#endif
+}
+
+inline float safe_fabs(const uint32_t v) {
+    return std::abs(int32_t(v));
```
Comment (khancyr) [inline libraries/AP_Math/definitions.h:137]: won't we risk overflow here ?

**Review by khancyr — COMMENTED** [review]
yep. Should be probably fine and checked by unittest

File: libraries/AP_Math/definitions.h (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_Math/definitions.h]: Thanks, fixed

File: libraries/AP_Math/definitions.h [hunk 2]
```diff
@@ -117,3 +117,30 @@ static const double WGS84_E = (sqrt(2 * WGS84_F - WGS84_F * WGS84_F));
...
+
+inline double safe_fabs(const double v) {
+#ifdef ALLOW_DOUBLE_MATH_FUNCTIONS
+    return fabs(v);
+#else
+    return fabsf(static_cast<float>(v));
+#endif
+}
+
+inline uint32_t safe_fabs(const uint32_t v) {
```
Comment (tridge) [inline libraries/AP_Math/definitions.h:136]: why are we taking an absolute value of an unsigned value? If we are doing that then this could break existing code by effectively removing a cast
for example:
  float x = safe_fabs(7U) / 2;
will give a different answer to:
 float x = fabsf(7U) / 2;

File: libraries/AP_Math/definitions.h [hunk 3]
```diff
@@ -117,3 +117,30 @@ static const double WGS84_E = (sqrt(2 * WGS84_F - WGS84_F * WGS84_F));
...
+    return v;
+}
+
+inline int32_t safe_fabs(const int32_t v) {
```
Comment (tridge) [inline libraries/AP_Math/definitions.h:140]: again, this removes a cast, potentially breaking code

File: libraries/AP_Math/definitions.h [hunk 4]
```diff
@@ -117,3 +117,30 @@ static const double WGS84_E = (sqrt(2 * WGS84_F - WGS84_F * WGS84_F));
...
+    return std::abs(v);
+}
+
+inline int16_t safe_fabs(const int16_t v) {
```
Comment (tridge) [inline libraries/AP_Math/definitions.h:144]: same problem

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
if the problem is the templates, we could either use type specialisation:
https://en.cppreference.com/w/cpp/language/template_specialization
or use intermediate floats for the epsilon test

**Comment by hendjoshsr71** [comment]
Since it was asked on the DevCall here is an example of the warning as seen in the CI MacOS build tests,

image

**Review by timtuxworth — COMMENTED** [review]
Apologies if I have the protocol wrong, I just wanted to suggest that since you have defined a new "is_zero()" function and most of the uses of safe_fabs() are to test for zero, perhaps it would make sense to use is_zero() instead of (safe_fabs(r1xr2) < FLT_EPSILON) in all those cases.

File: libraries/AP_Math/definitions.h (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_Math/definitions.h]: ok, I have taken your suggestion of using template specialization and so avoiding the problem

File: libraries/AP_Math/definitions.h (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_Math/definitions.h]: Ditto

File: libraries/AP_Math/definitions.h (hunk 4 above)
Comment (andyp1per) [reply libraries/AP_Math/definitions.h]: Ditto

**Comment by andyp1per** [comment]
Switched to template specialization - is much cleaner

File: libraries/AP_Math/AP_Math.h [hunk 5]
```diff
@@ -82,7 +82,11 @@ inline bool is_negative(const T fVal1) {
  * @brief: Check whether a double is zero
  */
 inline bool is_zero(const double fVal1) {
-    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
+#ifdef ALLOW_DOUBLE_MATH_FUNCTIONS
+    return (fabs(fVal1) < DBL_EPSILON);
```
Comment (tridge) [inline libraries/AP_Math/AP_Math.h:86]: this needs to use FLT_EPSILON, otherwise we end up with numbers that are not is_positiive(), not is_negative() and not is_zero()
it also could lead to some overflows when using a !is_zero() number in a single prec context

File: libraries/AP_Math/vector2.h [hunk 6]
```diff
@@ -276,6 +276,19 @@ struct Vector2
...
 };
 
+// check if all elements are zero
+template<> inline bool Vector2<float>::is_zero(void) const {
+    return (fabsf(x) < FLT_EPSILON) && (fabsf(y) < FLT_EPSILON);
+}
+
+template<> inline bool Vector2<double>::is_zero(void) const {
+#ifdef ALLOW_DOUBLE_MATH_FUNCTIONS
+    return (fabs(x) < DBL_EPSILON) && (fabs(y) < DBL_EPSILON);
```
Comment (tridge) [inline libraries/AP_Math/vector2.h:286]: needs to be FLT_EPSILON (see above comment)

File: libraries/AP_Math/vector3.h [hunk 7]
```diff
@@ -315,6 +315,25 @@ class Vector3
...
+// check if all elements are zero
+template<> inline bool Vector3<float>::is_zero(void) const {
+    return (fabsf(x) < FLT_EPSILON) &&
+            (fabsf(y) < FLT_EPSILON) &&
+            (fabsf(z) < FLT_EPSILON);
+}
+
+template<> inline bool Vector3<double>::is_zero(void) const {
+#ifdef ALLOW_DOUBLE_MATH_FUNCTIONS
+    return (fabs(x) < DBL_EPSILON) &&
```
Comment (tridge) [inline libraries/AP_Math/vector3.h:327]: needs to use FLT_EPSILON

**Review by tridge — CHANGES_REQUESTED** [review]
we need a unit test to ensure that values of FLT_EPSILON are considered zero for vector2, vector3

File: libraries/AP_Math/AP_Math.h (hunk 5 above)
Comment (timtuxworth) [reply libraries/AP_Math/AP_Math.h:86]: I'm pretty sure that would not be semantically correct. There is a reason why FLT_EPSILON and DBL_EPSILON exist, to give a value "so close to zero that you can't tell the difference base on the inaccuracies of floating point math". Just using FLT_EPSILON when you have a double means that values that are semantically not zero will be "zero". How is that ok? the <double> version of is_positive() and is_negative() should use DBL_EPSILON. To avoid the problem you raised @tridge - perhaps is_zero() should be defined as !is_positive() && !is_negative().

Comment (tridge) [reply libraries/AP_Math/AP_Math.h:86]: the thing is that we don't want "so close to zero that you can't tell the difference base on the inaccuracies of floating point math"
we're trying to avoid our controllers producing NaN and Inf values with complex equations. Going closer to the zero value with an equation that mixes single and double precision just makes it more likely we will blow up

Comment (timtuxworth) [reply libraries/AP_Math/AP_Math.h]: If the precision is not required/necessary or useful then perhaps then don't use double at all?

**Comment by timtuxworth** [comment]
I just pulled the latest code from master and got flooded with fabs() errors. Not just building SITL this time, now I get them building plane for the Pixhawk.

../../libraries/AP_Math/AP_Math.h:85:13: warning: absolute value function 'fabsf' given an argument of type 'const double' but has parameter of type 'float' which may cause truncation of value [-Wabsolute-value]
    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
            ^
../../libraries/AP_Math/AP_Math.h:85:13: note: use function 'std::abs' instead
    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
            ^~~~~
            std::abs
1 warning generated.

**Comment by timtuxworth** [comment]
I thought I messed up by building for a Pixhawk, so I did a waf clean, then configure for pixracer. Again - flooded with fabs() errors:

../../libraries/AP_Math/AP_Math.h:85:13: warning: absolute value function 'fabsf' given an argument of type 'const double' but has parameter of type 'float' which may cause truncation of value [-Wabsolute-value]
    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
            ^
../../libraries/AP_Math/AP_Math.h:85:13: note: use function 'std::abs' instead
    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
            ^~~~~
            std::abs
1 warning generated.

**Comment by timtuxworth** [comment]
I posted about some build errors, but it seems for some reason waf is building for SITL no matter what board I build for, so I don't think its anything to do with this issue. (I've added comments to [https://github.com/ArduPilot/ardupilot/issues/19853](https://github.com/ArduPilot/ardupilot/issues/19853)

File: libraries/AP_Math/vector3.h (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_Math/vector3.h]: Fixed

File: libraries/AP_Math/vector2.h (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_Math/vector2.h]: Fixed

**Comment by andyp1per** [comment]
@timtuxworth all builds fine on my branch which is rebased on master as of today

**Comment by timtuxworth** [comment]
> @timtuxworth all builds fine on my branch which is rebased on master as of today

I feel really silly and I have to apologise. I had used "./waf configure board=pixracer" to configure the build. In my defence this didn't throw an error and reported Success, but I was actually building for SITL again.

**Review by peterbarker — APPROVED** [review]

**Comment by tridge** [comment]
I'd like to see this flown on both copter and plane on real vehicles

**Comment by timtuxworth** [comment]
> I'd like to see this flown on both copter and plane on real vehicles

I can do it on a plane (Bixler 2 + PixRacer) - anything specific you would like to see?

**Comment by timtuxworth** [comment]
I just built this PR on my Mac and I'm getting these warnings:
./../libraries/AP_Math/AP_Math.h:85:13: warning: absolute value function 'fabsf' given an argument of type 'const double' but has parameter of type 'float' **which may cause truncation of value** [-Wabsolute-value]
    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
            ^
../../libraries/AP_Math/AP_Math.h:85:13: note: use function 'std::abs' instead
    return (fabsf(fVal1) < static_cast<double>(FLT_EPSILON));
            ^~~~~
            std::abs
1 warning generated.

**Comment by andyp1per** [comment]
> I just built this PR on my Mac (for SITL) and I'm getting these warnings (lots of them): ./../libraries/AP_Math/AP_Math.h:85:13: warning: absolute value function 'fabsf' given an argument of type 'const double' but has parameter of type 'float' **which may cause truncation of value** [-Wabsolute-value] return (fabsf(fVal1) < static_cast(FLT_EPSILON)); ^ ../../libraries/AP_Math/AP_Math.h:85:13: note: use function 'std::abs' instead return (fabsf(fVal1) < static_cast(FLT_EPSILON)); ^~~~~ std::abs 1 warning generated.

Ok, this shouldn't happen - let me investigate

**Comment by timtuxworth** [comment]
> > Ok, this shouldn't happen - let me investigate
> 
> @timtuxworth this builds fine for me on sitl on my mac - can you double check you have updated correctly?
> 
> The line numbers do not match your error - I think you have a bad update

I started from scratch. Pulled the latest master and cherry-picked the two commits from the PR then did a configure, clean and build. I'm enclosing a complete log of my session. Please let me know if I did anything wrong.
[ArduPlaneBuild2022-02-07.pdf](https://github.com/ArduPilot/ardupilot/files/8025309/ArduPlaneBuild2022-02-07.pdf)

**Comment by andyp1per** [comment]
@timtuxworth ah right now I see - you want *all* the warnings nuked. I wasn't trying to fix that, only the most egregious ones. That said the ones remaining in AP_Math are worth fixing, so I have added those in 😄

**Comment by andyp1per** [comment]
Flown successfully on a 5" copter - no issues. Loiter, Stabilize, Acro, RTL

**Comment by tridge** [comment]
tested on a plane

**Review by tridge — APPROVED** [review]
